### PR TITLE
feat(esp32): :sparkles: add esp32 support

### PR DESCRIPTION
**Pull request by Flaykz**

### CODE REVIEW

**Comment by sidoh** [comment]
This looks awesome! When you're ready I should be able to run the test suite

**Comment by Flaykz** [comment]
@sidoh, its ready for a review. i didnt had a esp8266 to test if there is any regression :/

**Comment by sidoh** [comment]
That's okay! I have an automated test suite I can run against both the esp32 and esp8266.

File: platformio.ini [hunk 1]
```diff
@@ -46,46 +47,84 @@ build_flags =
...
 ; -D MQTT_DEBUG
 ; -D MILIGHT_UDP_DEBUG
 ; -D STATE_DEBUG
 
+[esp8266]
+platform = espressif8266@~4
+lib_deps =
+  ${base.lib_deps_builtin}
+  ${base.lib_deps_external}
+  https://github.com/sidoh/WiFiManager.git#cmidgley
```
Comment (sidoh) [inline platformio.ini:61]: This should be `https://github.com/tzapu/WiFiManager.git#v2.0.16-rc.2`.

File: platformio.ini
Comment (sidoh) [inline platformio.ini:1]: I think we can avoid some repetition by using `extends`. Here's a diff:

```
> git --no-pager diff .\platformio.ini
diff --git a/platformio.ini b/platformio.ini
index 5f38d6c..f347b5a 100644
--- a/platformio.ini
+++ b/platformio.ini
@@ -22,9 +22,8 @@ default_envs =
 framework = arduino
 board_f_cpu = 160000000L
 lib_deps_builtin =
-  SPI
 lib_deps_external =
-  nrf24/RF24 @ ^1.4.9
+  nrf24/RF24@~1.3.2
   ArduinoJson
   PubSubClient@~2.8
   https://github.com/ratkins/RGBConverter.git#07010f2
@@ -54,13 +53,15 @@ build_flags =
 ; -D STATE_DEBUG

 [esp8266]
+extends = base
 platform = espressif8266@~4
 lib_deps =
   ${base.lib_deps_builtin}
   ${base.lib_deps_external}
-  https://github.com/sidoh/WiFiManager.git#cmidgley
+  https://github.com/tzapu/WiFiManager.git#v2.0.16-rc.2

 [esp32]
+extends = base
 platform = espressif32
 lib_deps =
   ${base.lib_deps_builtin}
@@ -69,57 +70,40 @@ lib_deps =
   luc-github/ESP32SSDP@^1.2.1

 [env:nodemcuv2]
-extends = base
-platform = ${esp8266.platform}
-framework = ${base.framework}
+extends = esp8266
 board = nodemcuv2
 build_flags = ${base.build_flags} -D FIRMWARE_VARIANT=nodemcuv2

 [env:d1_mini]
-extends = base
-platform = ${esp8266.platform}
-framework = ${base.framework}
+extends = esp8266
 board = d1_mini
 build_flags = ${base.build_flags} -D FIRMWARE_VARIANT=d1_mini

 [env:esp12]
-extends = base
-platform = ${esp8266.platform}
-framework = ${base.framework}
+extends = esp8266
 board = esp12e
 build_flags = ${base.build_flags} -D FIRMWARE_VARIANT=esp12

 [env:esp07]
-extends = base
-platform = ${esp8266.platform}
-framework = ${base.framework}
+extends = esp8266
 board = esp07
 build_flags = ${base.build_flags} -D FIRMWARE_VARIANT=esp07
 board_build.ldscript = eagle.flash.1m64.ld

 [env:huzzah]
-extends = base
-platform = ${esp8266.platform}
-framework = ${base.framework}
+extends = esp8266
 board = huzzah
 build_flags = ${base.build_flags} -D FIRMWARE_VARIANT=huzzah

 [env:d1_mini_pro]
-extends = base
-platform = ${esp8266.platform}
-framework = ${base.framework}
+extends = esp8266
 board = d1_mini_pro
 build_flags = ${base.build_flags} -D FIRMWARE_VARIANT=d1_mini_PRO

 [env:esp32doit-devkit-v1]
-extends = base
-platform = ${esp32.platform}
-framework = ${base.framework}
+extends = esp32
 board = esp32doit-devkit-v1
 build_flags = ${base.build_flags} -D FIRMWARE_VARIANT=esp32doit-devkit-v1
-extra_scripts = ${base.extra_scripts}
-lib_deps = ${esp32.lib_deps}
-test_ignore = ${base.test_ignore}

 [env:debug]
 extends = env:d1_mini
```

File: dist/global.h
Comment (sidoh) [inline dist/global.h:1]: can we put this in `lib`? `dist` is meant for compiled assets (like the webpage).

File: platformio.ini [hunk 2]
```diff
@@ -16,15 +16,16 @@ default_envs =
...
-board_build.ldscript = eagle.flash.4m1m.ld
-lib_deps =
-  https://github.com/tzapu/WiFiManager.git#v2.0.16-rc.2
-  RF24@~1.3.2
-  ArduinoJson@~6.21
+board_f_cpu = 160000000L
+lib_deps_builtin =
+  SPI
+lib_deps_external =
+  nrf24/RF24 @ ^1.4.9
```
Comment (sidoh) [inline platformio.ini:27]: I get compile errors for esp8266 using `nrf24/RF24` v1.4.9. If we need to be using 1.4.9 for esp32, let's split it into the platform-specific dependencies so we can keep the version at `~1.3.2` for ESP8266.

Comment (sidoh) [reply platformio.ini:27]: For reference this is the compile error:

```
.pio\libdeps\d1_mini\RF24/RF24_config.h:189:26: error: 'class HardwareSerial' has no member named 'Serial'
```

**Review by sidoh — COMMENTED** [review]
Looks good! I left a few comments -- need to make some minor changes to fix the esp8266 build.

Was able to verify that the integration test suite for ESP8266 is still working with these changes!

One additional request:

Can you update the README with the pin mappings for the ESP32 + nrf24l01? A new section around here is fine:

https://github.com/sidoh/esp8266_milight_hub?tab=readme-ov-file#nrf24l01

Doesn't need to be super detailed. Just the raw mapping is fine.

I haven't run the integration tests against ESP32 yet (will wait on the pin mappings so I'm not guessing), but if that works should be able to merge this pretty quickly!

**Comment by david-brouste** [comment]
Alright, I'll look at it after my workday :)

File: src/main.cpp [hunk 3]
```diff
@@ -367,7 +393,8 @@ void postConnectSetup() {
 
 void setup() {
   Serial.begin(9600);
-  String ssid = "ESP" + String(ESP.getChipId());
+  delay(1000);
```
[the remark this answers is not in the document]
Comment (Flaykz) [reply src/main.cpp:396]: I had put it for some tests, I forgot to remove it.

File: platformio.ini (hunk 2 above)
Comment (Flaykz) [reply platformio.ini:27]: My bad, better to keep 1.3.2 for esp32 too :)

**Comment by Flaykz** [comment]
@sidoh, thx for the review, changes made

**Comment by sidoh** [comment]
Got it working! Had to add this line to `platformio.ini`:

```
[esp32]
board_build.partitions = default.csv
```

but I think that's because the board I used was at some point flashed with a non-default partition table. Probably still worth including this since it seems like SPIFFS maybe doesn't work with non-default partition table.

I ran the integration tests. There were some failures, but I think only 4 of them are real. 

Here are the full results:

```
Failures:

  1) MQTT commands and state should respect the state update interval
     Failure/Error: expect(update_timestamp_gaps.length).to be >= 3

       expected: >= 3
            got:    2
     # ./spec/mqtt_spec.rb:216:in `block (3 levels) in <top (required)>'

  2) REST Server backup routes should preserve settings when creating a backup
     Failure/Error: res.value

     Net::HTTPServerException:
       400 "Bad Request"
     # ./lib/api_client.rb:119:in `block in request'
     # ./lib/api_client.rb:92:in `request'
     # ./lib/api_client.rb:144:in `upload_json'
     # ./lib/api_client.rb:152:in `block in upload_string_as_file'
     # ./lib/api_client.rb:149:in `upload_string_as_file'
     # ./spec/rest_spec.rb:366:in `block (3 levels) in <top (required)>'

  3) REST Server backup routes should override existing settings when restoring a backup
     Failure/Error: res.value

     Net::HTTPServerException:
       400 "Bad Request"
     # ./lib/api_client.rb:119:in `block in request'
     # ./lib/api_client.rb:92:in `request'
     # ./lib/api_client.rb:144:in `upload_json'
     # ./lib/api_client.rb:152:in `block in upload_string_as_file'
     # ./lib/api_client.rb:149:in `upload_string_as_file'
     # ./spec/rest_spec.rb:400:in `block (3 levels) in <top (required)>'

  4) UDP servers color and brightness commands should result in state changes
     Failure/Error:
       @client.get(topic) do |topic, message|
         ret_val = yield(topic, message)
         raise BreakListenLoopError if ret_val
       end

     Timeout::Error:
       execution expired
     # ./lib/mqtt_client.rb:90:in `block (2 levels) in on_message'
     # ./lib/mqtt_client.rb:89:in `block in on_message'

  5) UDP servers discovery should respond to v5 discovery
     Failure/Error: response, _ = @discovery_socket.recvfrom_nonblock(1024)

     IO::EAGAINWaitReadable:
       Resource temporarily unavailable - recvfrom(2) would block
     # ./spec/udp_spec.rb:127:in `block (3 levels) in <top (required)>'

  6) UDP servers discovery should respond to v6 discovery
     Failure/Error: expect(response.length).to      eq(3), "Should be a comma-separated list with three elements"
       Should be a comma-separated list with three elements
     # ./spec/udp_spec.rb:144:in `block (3 levels) in <top (required)>'

Finished in 9 minutes 16 seconds (files took 0.33943 seconds to load)
163 examples, 6 failures

Failed examples:

rspec ./spec/mqtt_spec.rb:181 # MQTT commands and state should respect the state update interval
rspec ./spec/rest_spec.rb:337 # REST Server backup routes should preserve settings when creating a backup
rspec ./spec/rest_spec.rb:383 # REST Server backup routes should override existing settings when restoring a backup
rspec ./spec/udp_spec.rb:88 # UDP servers color and brightness commands should result in state changes
rspec ./spec/udp_spec.rb:124 # UDP servers discovery should respond to v5 discovery
rspec ./spec/udp_spec.rb:135 # UDP servers discovery should respond to v6 discovery
```

I think we can ignore the MQTT error.

The backup route failures are real. Looks like the issue is we need to make `BACKUP_FILE` in `Settings.h` `/backup.bin` (instead of `_backup.bin`).

The first UDP failure looks like it was transient -- probably just a flaky test.

The UDP discovery failures seem real. Not the most important thing in the world -- the UDP server probably isn't used very much, but worth looking into. I'm out of time for tonight and will probably be a few days before I can look again, but suspect there's something extra we need to do for the ESP32 to allow receiving broadcast packets?

The ruby code that sends the broadcast packets is here in case it's helpful:

https://github.com/sidoh/esp8266_milight_hub/blob/master/test/remote/spec/udp_spec.rb#L119

**Comment by sidoh** [comment]
(I should mention that tons of tests pass -- the important stuff seems to be working just fine)

**Comment by sidoh** [comment]
Going to merge this into a branch in this repo so I can make the last few changes to get it over the line. tyvm for the PR <3

**Pull request merged**